src/Native/Unix/System.Native/pal_signal.c

```diff
@@ -108,6 +111,15 @@ static void* SignalHandlerLoop(void* arg)
             return NULL;
         }
 
+        if (signalCode == SIGCHLD || signalCode == SIGCONT || signalCode == SIGWINCH)
+        {
+            TerminalInvalidationCallback callback = g_terminalInvalidationCallback;
+            if (callback != NULL)
+            {
+                callback();
+            }
```

Comment:
Because there's asynchrony involved here, there could be an arbitrary period of time between which one of these signals arrives and we invoke the callback.

I wonder if we should instead consider an approach like what mono uses.  IIRC, mono has some static in native storing the cached position, and in some initialization code it hands the address of that field back to managed, which then caches that address.  To invalidate the cached position, the native code simply writes -1 or some such sentinel into that field, and the managed code uses that dereferenced address as its cache.

Would that work / be better here, or are there problems that introduces such that we'd be better off with the non-determinism of the async callback approach?

Reply:
(That said, there's already an inherent race condition in any code that's reading the position, doing something based on it, and then setting it, so maybe the asynchrony here is a red herring.)

Reply:
I'm going to have a closer look at the mono implementation. These signals maybe somewhat delayed. In the end, the cached window size and cursor position are invalidated. The 'version' field makes sure that happens even when there are on-going writes/getcursorposition operations.

Reply:
I had a look at the mono sources, but didn't find code that performs invalidation.

Reply:
> I had a look at the mono sources, but didn't find code that performs invalidation.

Yeah, I misspoke.  Mono actually has a bug here where the cursor position is not invalidated when the window resizes.

Here's the code that passes the native field's address out to the managed code:
https://github.com/mono/mono/blob/a80d44980cccef3306c23fe55da231e5b716083e/mono/metadata/console-unix.c#L454
The signal handler then just updates this field rather than calling back to managed, and the managed code reads that field via dereferencing the pointer it got:
https://github.com/mono/mono/blob/5852d1cf74257b06b30fd7ad23b9427a37343fe3/mcs/class/corlib/System/TermInfoDriver.cs#L706

Reply:
I hadn't noticed the `native_terminal_size` implementation. Since we're also invalidating the cursor position, your example/repro is working better.